go/upgrade/upgrade.go

```diff
@@ -279,8 +279,29 @@ func (u *upgradeManager) ConsensusUpgrade(privateCtx interface{}, currentEpoch b
 			if err := u.flushDescriptorLocked(); err != nil {
 				return err
 			}
-			u.shouldStop = true // Ensure we really stop before proceeding.
-			return api.ErrStopForUpgrade
+			// Check if we can proceed in place (e.g. without a restart).
+			u.shouldStop = func() bool {
+				if err := pu.Descriptor.EnsureCompatible(); err != nil {
+					// Not compatible, we must stop.
+					return true
+				}
+				// Check if the migration handler has a startup stage.
+				handler, err := migrations.GetHandler(pu.Descriptor.Handler)
+				if err != nil {
+					// Handler not available, we must stop.
```

Comment:
Should we log this, as this is not expected?
I would also log "skipping startup upgrade".

Reply:
It could be expected as you can have an upgrade that does not change the protocol versions but only adds an upgrade handler. Will add a log entry for skipping startup upgrade though.